src/aggregate.js

```diff
@@ -0,0 +1,230 @@
+import * as API from './api.js'
+import * as Hybrid from './hybrid.js'
+import * as Segment from './segment.js'
+import * as Index from './index.js'
+import * as Piece from './piece.js'
+import * as Node from './node.js'
+import { log2Ceil } from './uint64.js'
+import { indexAreaStart } from './inclusion.js'
+
+const NodeSize = BigInt(Node.Size)
+const EntrySize = Number(Index.EntrySize)
+export const MAX_CAPACITY = 2n ** BigInt(Hybrid.MAX_LOG2_LEAFS) * NodeSize
+
+/**
+ * Our default aggregate size is 32GiB. 16GiB might be an option smaller pieces
+ *
+ * but our current
+ * average rate is 30 GiB per hour
+ */
+export const DEFAULT_DEAL_SIZE = Piece.PaddedSize.from(2n ** 35n) // 32 GiB
+
+export const { PaddedSize, UnpaddedSize } = Piece
+
+/**
+ * @param {object} [options]
+ * @param {API.PaddedPieceSize} [options.size] - Size of the aggregate in
+ * (fr32 padded) bytes. If omitted default to 32 GiB which is a good
+ * default because deals 8GiB and smaller are too expensive for service
+ * providers to bother. The 16GiB deals may be an option, however our
+ * current average rate of 30 GiB per hour suggests that 32 GiB is a
+ * better default.
+ */
+export const createBuilder = ({ size = DEFAULT_DEAL_SIZE } = {}) =>
+  new AggregateBuilder({ size })
+
+/**
+ * @param {object} options
+ * @param {API.PieceInfo[]} options.pieces - Pieces to add to the aggregate
+ * @param {API.PaddedPieceSize} [options.size] - Size of the aggregate in
+ * (fr32 padded) bytes. If omitted default to 32 GiB
+ */
+export const build = ({ pieces, size = DEFAULT_DEAL_SIZE }) => {
+  const builder = createBuilder({ size })
+
+  for (const piece of pieces) {
+    builder.write(piece)
+  }
+
+  return builder.close()
+}
+
+class AggregateBuilder {
+  /**
+   * @param {object} source
+   * @param {API.PaddedPieceSize} source.size
+   * @param {API.uint64} [source.offset]
+   * @param {API.MerkleTreeNodeSource[]} [source.parts]
+   * @param {number} [source.limit]
+   */
+  constructor({
+    size,
```

Comment:
Should we have a minimum size that would fail to close? We can set the defaults we know today but still tweak as needed

Reply:
What would that minimum size imply, amount of pieces you need to add ? I would rather have a field reporting how much space there is left to fill instead personally.

Also note that aggregate size limits amount of pieces it may contain, so it could be that you have not met the min size utilization, but you did maxed out on number of pieces

Reply:
I was thinking about the minimum requirement of Spade `1+127*(1<<27)` or `15.875GiB` to be the default value. But yes, the maxed number of pieces with it adds some complexity

Reply:
Oh you mean actual aggregate size as opposed to space utilization which is what I though you were suggesting. We could, but it kind of feels wrong to enforce spade requirements at the library level as someone may want to use this for something else. It's also nice to be able to test the library on smaller sizes.

I'll leave it as is, if we decide to do it we can do as followup.